### PR TITLE
Regenerate checksum when chunks are to be sliced

### PR DESCRIPTION
**Pull request by kuenishi**

### CODE REVIEW

File: src/machi_file_proxy.erl [hunk 1]
```diff
@@ -664,3 +659,26 @@ do_write(FHd, CsumTable, Filename, TaggedCsum, Offset, Size, Data) ->
...
+    NewH = {F, LeftPos, NewChunk, NewChecksum},
+    slice_both_side([NewH|L], LeftPos, RightPos);
+slice_both_side(Chunks, LeftPos, RightPos) when LeftPos =< RightPos ->
+    %% TODO: optimize
+    [{F, Offset, Chunk, _Csum}|L] = lists:reverse(Chunks),
+    Size = iolist_size(Chunk),
+    if RightPos < Offset + Size ->
+            NewSize = RightPos - Offset,
+            <<NewChunk:NewSize/binary, _/binary>> = Chunk,
+            NewChecksum = machi_util:make_tagged_csum(client_sha, Chunk),
```
Comment (kuenishi) [inline src/machi_file_proxy.erl]: Well, so this should be `server_sha`.

Comment (slfritchie) [inline src/machi_file_proxy.erl]: Hrm, perhaps ?CSUM_TAG_SERVER_REGEN_SHA.  To be honest, I don't have 100% solid definitions for each of the checksum types in `include/machi.hrl.  But we probably should create a good definition for each one!

When I first wrote them, my feeling was that ?CSUM_TAG_SERVER_SHA would be created by the server when the client was stupid and gave the server a ?CSUM_TAG_NONE.  Perhaps the data was corrupted by the network or some other place, silly client!  So ?CSUM_TAG_SERVER_SHA is a substitute for an original, client-generated checksum.

Then ?CSUM_TAG_SERVER_REGEN_SHA would be used whenever the server had verified an original client_sha or server_sha but had to recalculate due to a trim or a read that is not at an original chunk offset + size.

Comment (slfritchie) [inline src/machi_file_proxy.erl]: If the above ^^ sounds good, or if it needs some minor changes, then ... after fixing, do you have a suggestion for where we should make these definitions easy to find?  machi.hrl, or someplace else?  This GH ticket is probably not a good long-term place.

**Comment by kuenishi** [comment]
> This GH ticket is probably not a good long-term place.

Exactly. I tried to describe the definition in a `@doc` manner in `machi_util.erl` and `machi_dt.erl` in the last commit push 0e4ae81 . As well as I defined macros of atoms to avoid typo bug (I made so many bugs with atom typo).

File: src/machi_dt.erl [hunk 2]
```diff
@@ -44,10 +44,16 @@
 -type projection()      :: #projection_v1{}.
 -type projection_type() :: 'public' | 'private'.
 
+%% @doc Tags that stand for how that checksum was generated. See
+%% machi_util:make_tagged_csum/{1,2} for further documentation and
+%% implementationo.
+-type csum_tag()    :: none | client_sha | server_sha | server_regen_sha.
```
Comment (jadeallenx) [inline src/machi_dt.erl:50]: Small typo here. `implementation`

Comment (kuenishi) [inline src/machi_dt.erl:50]: Thanks! Fixed typo with the following commit.

**Comment by slfritchie** [comment]
Hm.

```
==> machi (dialyzer)
machi_file_proxy.erl:665: The pattern <[], _, _> can never match the type <[any(),...],non_neg_integer(),non_neg_integer()>
```

**Comment by slfritchie** [comment]
Two compiler & Dialyzer warnings fixed. +1, thanks!

**Comment by kuenishi** [comment]
Oh, sorry.  I'll keep running dialyzer next time.

**Pull request merged**